src/libopenrave/kinbody.cpp

```diff
@@ -5348,56 +5446,141 @@ const std::vector<int>& KinBody::GetNonAdjacentLinks(int adjacentoptions) const
                 }
             }
         }
-private:
+    private:
         KinBodyConstPtr _pbody;
         std::vector<Transform> vcurtrans;
         std::vector<dReal> _vdoflastsetvalues;
     };
 
-    CHECK_INTERNAL_COMPUTATION;
-    if( _nNonAdjacentLinkCache & 0x80000000 ) {
-        // Check for colliding link pairs given the initial pose _vInitialLinkTransformations
-        // this is actually weird, we need to call the individual link collisions on a const body. in order to pull this off, we need to be very careful with the body state.
-        TransformsSaver saver(shared_kinbody_const());
-        CollisionCheckerBasePtr collisionchecker = !!_selfcollisionchecker ? _selfcollisionchecker : GetEnv()->GetCollisionChecker();
-        CollisionOptionsStateSaver colsaver(collisionchecker,0); // have to reset the collision options
+    _ResizeVectorFor2DTable(adjacentLinkFlags, _veclinks.size());  // Fills with 0
+    if( _vNonSelfCollidingPositionConfigurationsAndLinkTransformations.empty() ) {
+        return;
+    }
+
+    // Check for colliding link pairs given _vNonSelfCollidingPositionConfigurationsAndLinkTransformations
+    // this is actually weird, we need to call the individual link collisions on a const body. in order to pull this off, we need to be very careful with the body state.
+    TransformsSaver saver(shared_kinbody_const());
+    CollisionCheckerBasePtr collisionchecker = !!_selfcollisionchecker ? _selfcollisionchecker : GetEnv()->GetCollisionChecker();
+    CollisionOptionsStateSaver colsaver(collisionchecker, 0); // have to reset the collision options
+
+    std::vector<bool> dofValueIsDeterminableFlags(GetDOF());  // Used to determine which link pairs can be evaluated for adjacency
+    for( const PositionConfigurationAndLinkTransformations& positionConfigurationAndLinkTransformations : _vNonSelfCollidingPositionConfigurationsAndLinkTransformations ){
+        // Applies the non-self-colliding position configuration
+        OPENRAVE_ASSERT_OP(_veclinks.size(), ==, positionConfigurationAndLinkTransformations.second.size());
         for(size_t i = 0; i < _veclinks.size(); ++i) {
-            boost::static_pointer_cast<Link>(_veclinks[i])->_info._t = _vInitialLinkTransformations.at(i);
+            boost::static_pointer_cast<Link>(_veclinks[i])->_info._t = positionConfigurationAndLinkTransformations.second.at(i);
         }
         _nUpdateStampId++; // because transforms were modified
-        _vNonAdjacentLinks[0].resize(0);
 
-        for(size_t ind0 = 0; ind0 < _veclinks.size(); ++ind0) {
-            for(size_t ind1 = ind0+1; ind1 < _veclinks.size(); ++ind1) {
-                const bool bAdjacent = AreAdjacentLinks(ind0, ind1);
-                if(!bAdjacent && !collisionchecker->CheckCollision(LinkConstPtr(_veclinks[ind0]), LinkConstPtr(_veclinks[ind1])) ) {
-                    _vNonAdjacentLinks[0].push_back(ind0|(ind1<<16));
+        // Constructs initial dofValueIsDeterminableFlags
```

Comment:
Could you add the reason we need to do this to these comments?

Reply:
Added more comments.